src/clients/UBAClient/UBAClientUtilities.ts

```diff
@@ -232,11 +241,150 @@ export function getMostRecentBundles(
  * This can be used to validate a deposit that is much older than spoke pool client's lookback.
  * @param flow
  */
-// async function getBundleStateForFlow(flow: UbaFlow): ModifiedUBAFlow {
-//   // 1. First try to fetch flow information from existing bundle state saved in client memory.
-//   // 2. If not available, then we need to load additional data from the RPC.
-//   throw new Error("Unimplemented");
-// }
+export async function getModifiedFlow(
+  chainId: number,
+  flow: UbaFlow,
+  hubPoolClient: HubPoolClient,
+  spokePoolClients: SpokePoolClients
+): Promise<ModifiedUBAFlow> {
+  let tokenSymbol: string | undefined;
+  if (isUbaInflow(flow)) {
+    if (chainId !== flow.originChainId) {
+      throw new Error(`ChainId ${chainId} should be deposit.originChainId`);
+    }
+    tokenSymbol = hubPoolClient.getTokenInfo(flow.originChainId, flow.originToken)?.symbol;
+  } else if (outflowIsFill(flow as UbaOutflow)) {
+    if (chainId !== flow.destinationChainId) {
+      throw new Error(`ChainId ${chainId} should be fill.destinationChainId`);
+    }
+    tokenSymbol = hubPoolClient.getTokenInfo(flow.destinationChainId, (flow as FillWithBlock).destinationToken)?.symbol;
+  } else if (chainId !== (flow as RefundRequestWithBlock).repaymentChainId) {
+    if (chainId !== flow.repaymentChainId) {
+      throw new Error(`ChainId ${chainId} should be refund.repaymentChainId`);
+    }
+    tokenSymbol = hubPoolClient.getTokenInfo(
+      flow.repaymentChainId,
+      (flow as RefundRequestWithBlock).refundToken
+    )?.symbol;
+  }
+  if (!tokenSymbol) {
+    throw new Error(`Could not find token symbol for chain ${chainId} and flow ${JSON.stringify(flow)}`);
+  }
+
+  // Load bundle ranges before flow and until after it:
+  const bundleRanges = Object.fromEntries(
+    Object.keys(spokePoolClients).map((_chainId) => {
+      const bundles = getMostRecentBundleBlockRanges(
+        Number(_chainId),
+        // 3 seems like a reasonable number of bundles to lookback to guarantee we can validate all events in the
+        // most recent one.
+        3,
+        flow.quoteBlockNumber,
+        hubPoolClient,
+        spokePoolClients
+      );
+      // Make bundle.end cover from the block range until the flow.block so we know the running balance right before
+      // the flow
+      bundles[bundles.length - 1].end = flow.blockNumber;
+      return [_chainId, bundles];
+    })
+  );
```

Comment:
I think we can use mapAsync here to get additional speedups as each bundle is independent of one another.

Reply:
There are no blocking calls in here - how would `mapAsync` help to speed it up?

Reply:
Oh goodness - I thought I saw an async call. I'll resolve this comment now